File: smartadapter/src/main/java/smartadapter/internal/mapper/ViewEventMapper.java
package smartadapter.internal.mapper;

/*
 * Created by Manne Öhlund on 2019-08-01.
 * Copyright (c) All rights reserved.
 */

import android.util.Log;
import android.util.SparseArray;
import android.view.View;

import androidx.annotation.NonNull;

import java.util.HashMap;

import smartadapter.R;
import smartadapter.SmartRecyclerAdapter;
import smartadapter.listener.OnViewEventListener;
import smartadapter.viewholder.SmartViewEventListenerHolder;
import smartadapter.viewholder.SmartViewHolder;

import static smartadapter.internal.utils.ViewEventValidator.isViewEventIdValid;

/**
 * Maps and binds view events with view holder target views with {@link ViewEventBinderProvider}.
 */
public class ViewEventMapper {

    private HashMap<Class<? extends SmartViewHolder>, SparseArray<SparseArray<OnViewEventListener>>> viewEventListenerMap = new HashMap<>();
    private ViewEventBinderProvider viewEventListenerMapperProvider = new ViewEventBinderProvider();

    /**
     * Adds {@link OnViewEventListener} to the {@link SmartRecyclerAdapter}.
     * The adapter will then automatically map the {@link OnViewEventListener} to the target view holder class with {@link OnViewEventListener#getViewHolderType()},
     * set the viewEventListener on the right View with viewId using {@link OnViewEventListener#getViewId()}.
     *
     * @see OnViewEventListener#getViewEventId() for predefined eventds.
     *
     * @param viewEventListener target OnViewEventListener
     */
    public final void addViewEventListener(OnViewEventListener viewEventListener) {
        if (!isViewEventIdValid(viewEventListener.getViewEventId()))
            throw new RuntimeException(String.format("Invalid view event id (%d) for ViewHolder (%s)", viewEventListener.getViewEventId(), viewEventListener.getViewHolderType()));

        SparseArray<SparseArray<OnViewEventListener>> mapper;
        if ((mapper = viewEventListenerMap.get(viewEventListener.getViewHolderType())) == null) {
            mapper = new SparseArray<>();
        }
        if (mapper.indexOfKey(viewEventListener.getViewId()) < 0) {
            final SparseArray<OnViewEventListener> viewEventAndListenerMap = new SparseArray<>();
            viewEventAndListenerMap.put(viewEventListener.getViewEventId(), viewEventListener);
            mapper.put(viewEventListener.getViewId(), viewEventAndListenerMap);
        }
        mapper.get(viewEventListener.getViewId()).put(viewEventListener.getViewEventId(), viewEventListener);
        this.viewEventListenerMap.put(viewEventListener.getViewHolderType(), mapper);
    }

    public HashMap<Class<? extends SmartViewHolder>, SparseArray<SparseArray<OnViewEventListener>>> getViewEventListenerMap() {
        return viewEventListenerMap;
    }

    public void mapViewActionWith(@NonNull SmartViewHolder smartViewHolder) {
        mapViewActionWith(smartViewHolder, viewEventListenerMap.get(SmartViewHolder.class));
        mapViewActionWith(smartViewHolder, viewEventListenerMap.get(smartViewHolder.getClass()));
    }

    private  void mapViewActionWith(@NonNull SmartViewHolder smartViewHolder, SparseArray<SparseArray<OnViewEventListener>> viewIdActionIdMap) {
        if (viewIdActionIdMap != null) {
            for (int i = 0; i < viewIdActionIdMap.size(); i++) {
                SparseArray<OnViewEventListener> eventIdAndListener = viewIdActionIdMap.valueAt(i);
                for (int j = 0; j < eventIdAndListener.size(); j++) {
                    int viewId = viewIdActionIdMap.keyAt(i);
                    int viewEventId = eventIdAndListener.keyAt(j);
                    OnViewEventListener viewActionListener = eventIdAndListener.valueAt(j);

                    if (viewId == R.id.undefined &&
                            viewEventId == R.id.undefined) {
                        if (SmartViewEventListenerHolder.class.isAssignableFrom(smartViewHolder.getClass())) {
                            ((SmartViewEventListenerHolder)smartViewHolder).setOnViewEventListener(viewActionListener);
                        } else {
                            Log.e(ViewEventMapper.class.getName(), String.format(
                                    "Don't forget that '%s' needs to implement '%s' in order to receive the events",
                                    smartViewHolder.getClass().getName(),
                                    SmartViewEventListenerHolder.class.getName()));
                        }
                    }

                    View targetView = smartViewHolder.itemView;
                    if (viewId != R.id.undefined) {
                        targetView = smartViewHolder.itemView.findViewById(viewId);
                    }

                    viewEventListenerMapperProvider.bind(smartViewHolder, targetView, viewActionListener, viewEventId);
                }
            }
        }
    }
}

File: smartadapter/src/main/java/smartadapter/internal/mapper/ViewEventMapper.kt
package smartadapter.internal.mapper

/*
 * Created by Manne Öhlund on 2019-08-01.
 * Copyright (c) All rights reserved.
 */

import android.util.SparseArray
import io.github.manneohlund.smartrecycleradapter.R
import smartadapter.SmartRecyclerAdapter
import smartadapter.SmartViewHolderType
import smartadapter.listener.OnViewEventListener
import smartadapter.listener.ViewEvent
import smartadapter.viewholder.SmartViewHolder
import smartadapter.viewholder.ViewEventListenerHolder
import java.util.HashMap

/**
 * Maps and binds view events with view holder target views with [ViewEventBinderProvider].
 */
class ViewEventMapper() {

    lateinit var smartRecyclerAdapter: SmartRecyclerAdapter
    val viewEventListenerMap = HashMap<SmartViewHolderType, SparseArray<SparseArray<OnViewEventListener<*>>>>()
    private val viewEventListenerMapperProvider = ViewEventBinderProvider()

    /**
     * Adds [OnViewEventListener] to the [SmartRecyclerAdapter].
     * The adapter will then automatically map the [OnViewEventListener] to the target view holder class with [OnViewEventListener.viewHolderType],
     * set the viewEventListener on the right View with viewId using [OnViewEventListener.viewId].
     *
     * @see OnViewEventListener.viewEventId
     * @param viewEventListener target OnViewEventListener
     */
    fun addViewEventListener(viewEventListener: OnViewEventListener<*>) {
        var mapper: SparseArray<SparseArray<OnViewEventListener<*>>>?
        mapper = viewEventListenerMap[viewEventListener.viewHolderType]
        if (mapper == null) {
            mapper = SparseArray()
        }
        if (mapper.indexOfKey(viewEventListener.viewId) < 0) {
            val viewEventAndListenerMap = SparseArray<OnViewEventListener<*>>()
            viewEventAndListenerMap.put(viewEventListenerMapperProvider.resolveViewEventId(viewEventListener), viewEventListener)
            mapper.put(viewEventListener.viewId, viewEventAndListenerMap)
        }
        mapper.get(viewEventListener.viewId).put(viewEventListenerMapperProvider.resolveViewEventId(viewEventListener), viewEventListener)
        this.viewEventListenerMap[viewEventListener.viewHolderType] = mapper
    }

    fun mapViewEventWith(smartViewHolder: SmartViewHolder<*>) {
        mapViewEventWith(smartViewHolder, viewEventListenerMap[SmartViewHolder::class])

        for ((key, value) in viewEventListenerMap) {
            val isNotOfSmartViewHolder = key != SmartViewHolder::class
            val isTargetClass = key.java.isAssignableFrom(smartViewHolder::class.java)
            if (isNotOfSmartViewHolder && isTargetClass) {
                mapViewEventWith(smartViewHolder, value)
            }
        }
    }

    private fun mapViewEventWith(
        smartViewHolder: SmartViewHolder<*>,
        viewIdViewEventIdMap: SparseArray<SparseArray<OnViewEventListener<*>>>?
    ) {
        if (viewIdViewEventIdMap != null) {
            for (i in 0 until viewIdViewEventIdMap.size()) {
                val eventIdAndListener = viewIdViewEventIdMap.valueAt(i)
                for (j in 0 until eventIdAndListener.size()) {
                    val viewId = viewIdViewEventIdMap.keyAt(i)
                    val viewEventId = eventIdAndListener.keyAt(j)
                    val viewEventListener = eventIdAndListener.valueAt(j)

                    if (viewId == R.id.undefined && smartViewHolder is ViewEventListenerHolder<*> && viewEventListener.listener is ViewEvent) {
                        (smartViewHolder as? ViewEventListenerHolder<ViewEvent>)?.let { viewEventListenerHolder ->
                            (viewEventListener.listener as? ViewEvent)?.let {
                                viewEventListenerHolder.viewEventListener = it
                            }
                        }
                    } else {
                        var targetView = smartViewHolder.itemView
                        if (viewId != R.id.undefined) {
                            targetView = smartViewHolder.itemView.findViewById(viewId)
                        }

                        viewEventListenerMapperProvider.bind(smartRecyclerAdapter, smartViewHolder, targetView, viewEventListener, viewEventId)
                    }
                }
            }
        }
    }
}
